docs/isl-2-0/spec.md

```diff
@@ -0,0 +1,673 @@
+```ion
+type::{
+  name: bar,
+  type: { one_of:[$null, int] }
+}
+```
+" %}
+
+### Variably-Occurring Type References
+
+Some constraints allow a quantity to be specified along with a type reference.
+The quantity is given by the `occurs` field, and the field value can be one of `optional`, `required`, a non-negative integer, or an integer range.
+
+{% include grammar-element.md productions="variably_occurring_type_reference,occurs" %}
+
+While it is valid for a constraint to be repeated, `occurs` is not a constraint and may not appear more than once in any type definition.
+The field `occurs` is never allowed in a `NAMED_TYPE_DEFINITION`.
+
+The `occurs` field is optional.
+Each constraint that accepts a variably-occurring type reference specifies what the default `occurs` value is for that constraint. 
+
+# Constraints
+
+Constraints are the fundamental building blocks of types.
+Each constraint is a rule that restricts the values that are valid for the type.
+
+The order of constraints does not matter.
+Constraints can be repeated, and when they are, both instances of the constraint will apply.
+
+{% comment %}{% include example.md title="Repeated constraints" code_file="examples/placeholder.isl" %}{% endcomment %}
+
+Generally speaking, constraints must reject null values as invalid.
+For example, the precision and scale constraints must reject a null value, as `null` does not have a precision or scale to evaluate.
+The `fields` constraint must reject `null.struct`, as `null.struct` doesn't have any fields.
+Similar reasoning applies to the expected handling of null values by most constraints.
+The `contains`, `annotations`, `valid_values`, and type-algebra constraints are exceptions to this, as these constraints may be defined such that a null value is valid.
+
+### Ranges
+
+{% include grammar-element.md productions="range_int,range_number,range_timestamp,range_timestamp_precision,number,exclusivity" %}
+
+Some constraints can be defined by a range.
+A range is represented by a list annotated with `range`, and containing two values, in order: the minimum and maximum ends of the range.
+The default behavior is for both ends of the range to be *inclusive*; 
+if *exclusive* behavior is desired, the minimum or maximum (or both) values shall be annotated with `exclusive`.
+If the minimum or maximum end of a range is to be unspecified, this shall be represented by the symbols `min` or `max`, respectively; 
+the `exclusive`annotation is not applicable when the symbols `min` or `max` are specified.
+A range may not contain both `min` and `max`.
+
+All ranges have a type.
+The type of the range is the same as that of the minimum and/or maximum values specified in the range list. 
+If both a minimum and maximum values are specified (i.e. `min` and `max` are not used), then both of those values must be of the same type. 
+(For example, `range::[1995-12-06T, 55.4]` mixes values of the timestamp and number types, and therefore is not a valid range.)
```

Comment:
Reading this example, I wondered how we distinguished between `[int, int]` ranges and `[number, number]`, but I guess it's precedence in the grammar? (If a range is made of ints it will match `[int, int]` first?)

Reply:
It depends on where it is being used.
- For `valid_values` a range should always be parsed as `[number, number]` or `[timestamp, timestamp]`.
- For `timestamp_precision` it must always be a timestamp precision range. 
- For every other constraint that uses a range (as well as `occurs`) the range must be `[int, int]`.